File: ml-api/app/v2_contracts.py
"""Additive Version 2 domain contracts.

This module is deliberately separate from the active V1 schemas.  Importing it
must not change V1 routing, persistence, authorization, or model behavior.
"""

from __future__ import annotations

from collections.abc import Mapping
from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field

V1_SCHEMA_VERSION = "v1"
V1_WORKFLOW_VERSION = "v1"
V1_TAXONOMY_VERSION = "v1"
V2_SCHEMA_VERSION = "v2"
V2_WORKFLOW_VERSION = "v2"
V2_TAXONOMY_VERSION = "v2"

V2_CATEGORY_IDS = (
    "account_branch_services",
    "cards_atm_pos",
    "mobile_internet_banking",
    "transfers_payments_remittance",
    "loans_credit",
    "fraud_scam_unauthorized",
    "kyc_verification_restrictions",
    "general_complaints",
)
V2_LIFECYCLE_STATUSES = (
    "received",
    "queued",
    "assigned",
    "under_investigation",
    "additional_information_required",
    "action_in_progress",
    "resolved",
    "closed",
    "reopened",
)
STAFF_EMPLOYMENT_STATES = (
    "invited",
    "active",
    "on_leave",
    "suspended",
    "departed",
    "archived",
)
STAFF_AVAILABILITY_STATES = ("available", "busy", "offline", "unavailable")

V2CategoryId = Literal[
    "account_branch_services",
    "cards_atm_pos",
    "mobile_internet_banking",
    "transfers_payments_remittance",
    "loans_credit",
    "fraud_scam_unauthorized",
    "kyc_verification_restrictions",
    "general_complaints",
]
V2LifecycleStatus = Literal[
    "received",
    "queued",
    "assigned",
    "under_investigation",
    "additional_information_required",
    "action_in_progress",
    "resolved",
    "closed",
    "reopened",
]
StaffEmploymentState = Literal[
    "invited", "active", "on_leave", "suspended", "departed", "archived"
]
StaffAvailabilityState = Literal["available", "busy", "offline", "unavailable"]

V1_TICKET_STATUSES = frozenset(
    {"submitted", "triaged", "in_progress", "awaiting_customer", "resolved", "closed"}
)
V1_CATEGORY_IDS = frozenset(
    {
        "transfer_payment",
        "account_support",
        "card_atm",
        "fraud_security",
        "loan_credit",
        "general_support",
    }
)


class V2ContractError(ValueError):
    """Raised when a V2 contract is unknown, malformed, or unsafe to read."""


class V2TicketContract(BaseModel):
    """Strict, additive V2 ticket contract used before future persistence."""

    model_config = ConfigDict(
        extra="forbid", populate_by_name=False, str_strip_whitespace=False
    )

    schema_version: Literal["v2"] = Field(alias="schemaVersion")
    workflow_version: Literal["v2"] = Field(alias="workflowVersion")
    taxonomy_version: Literal["v2"] = Field(alias="taxonomyVersion")
    category_id: V2CategoryId = Field(alias="categoryId")
    status: V2LifecycleStatus


class V2StaffProfileContract(BaseModel):
    """Strict V2 Staff lifecycle fields; no authentication behavior is changed."""

    model_config = ConfigDict(extra="forbid", populate_by_name=False)

    schema_version: Literal["v2"] = Field(alias="schemaVersion")
    employment_state: StaffEmploymentState = Field(alias="employmentState")
    availability: StaffAvailabilityState


class TicketVersionInfo(BaseModel):
    """Safe reader result; it never converts V1 values into V2 values."""

    model_config = ConfigDict(extra="forbid")

    schema_version: Literal["v1", "v2"] = Field(alias="schemaVersion")
    workflow_version: Literal["v1", "v2"] = Field(alias="workflowVersion")
    taxonomy_version: Literal["v1", "v2"] = Field(alias="taxonomyVersion")


def _raw_key_present(document: Mapping[str, Any]) -> bool:
    """Reject snake_case aliases at the V2 boundary."""

    return any("_" in key for key in document if isinstance(key, str))


def read_ticket_version(document: Mapping[str, Any]) -> TicketVersionInfo:
    """Read version metadata without rewriting or reinterpreting a ticket.

    A document without version metadata is treated as V1 only when its
    historical status and department fields are recognizable. Partial or
    unknown metadata fails closed.
    """

    if not isinstance(document, Mapping) or _raw_key_present(document):
        raise V2ContractError("ticket version metadata is not safely readable")
    version_keys = ("schemaVersion", "workflowVersion", "taxonomyVersion")
    present = [key in document for key in version_keys]
    if not any(present):
        status = document.get("status")
        department = document.get("departmentId")
        if status not in V1_TICKET_STATUSES or (
            department is not None and department not in V1_CATEGORY_IDS
        ):
            raise V2ContractError("missing version metadata is not proven V1")
        return TicketVersionInfo(
            schemaVersion=V1_SCHEMA_VERSION,
            workflowVersion=V1_WORKFLOW_VERSION,
            taxonomyVersion=V1_TAXONOMY_VERSION,
        )
    if not all(present):
        raise V2ContractError("version metadata is incomplete")
    values = tuple(document.get(key) for key in version_keys)
    if values == (V1_SCHEMA_VERSION, V1_WORKFLOW_VERSION, V1_TAXONOMY_VERSION):
        return TicketVersionInfo(
            schemaVersion=V1_SCHEMA_VERSION,
            workflowVersion=V1_WORKFLOW_VERSION,
            taxonomyVersion=V1_TAXONOMY_VERSION,
        )
    if values == (V2_SCHEMA_VERSION, V2_WORKFLOW_VERSION, V2_TAXONOMY_VERSION):
        return TicketVersionInfo(
            schemaVersion=V2_SCHEMA_VERSION,
            workflowVersion=V2_WORKFLOW_VERSION,
            taxonomyVersion=V2_TAXONOMY_VERSION,
        )
    raise V2ContractError("ticket version is unknown or inconsistent")


def validate_v2_category(category_id: Any) -> V2CategoryId:
    if category_id not in V2_CATEGORY_IDS:
        raise V2ContractError("unknown V2 category")
    return category_id  # type: ignore[return-value]


def validate_v2_lifecycle_status(status: Any) -> V2LifecycleStatus:
    if status not in V2_LIFECYCLE_STATUSES:
        raise V2ContractError("unknown V2 lifecycle status")
    return status  # type: ignore[return-value]


def validate_staff_profile_states(
    *, employment_state: Any, availability: Any
) -> tuple[StaffEmploymentState, StaffAvailabilityState]:
    if employment_state not in STAFF_EMPLOYMENT_STATES:
        raise V2ContractError("unknown Staff employment state")
    if availability not in STAFF_AVAILABILITY_STATES:
        raise V2ContractError("unknown Staff availability state")
    return employment_state, availability  # type: ignore[return-value]

File: frontend/src/lib/v2-contracts.ts
import type { Locale } from "@/lib/i18n";

export const v2DepartmentIds = [
  "account_branch_services",
  "cards_atm_pos",
  "mobile_internet_banking",
  "transfers_payments_remittance",
  "loans_credit",
  "fraud_scam_unauthorized",
  "kyc_verification_restrictions",
  "general_complaints",
] as const;

export type V2DepartmentId = (typeof v2DepartmentIds)[number];

export const v2LifecycleStatuses = [
  "received",
  "queued",
  "assigned",
  "under_investigation",
  "additional_information_required",
  "action_in_progress",
  "resolved",
  "closed",
  "reopened",
] as const;

export type V2LifecycleStatus = (typeof v2LifecycleStatuses)[number];

export const staffEmploymentStates = [
  "invited",
  "active",
  "on_leave",
  "suspended",
  "departed",
  "archived",
] as const;

export type StaffEmploymentState = (typeof staffEmploymentStates)[number];

export const staffAvailabilityStates = ["available", "busy", "offline", "unavailable"] as const;

export type StaffAvailabilityState = (typeof staffAvailabilityStates)[number];

export type TicketVersion = "v1" | "v2";

export type V2TicketContract = {
  schemaVersion: "v2";
  workflowVersion: "v2";
  taxonomyVersion: "v2";
  categoryId: V2DepartmentId;
  status: V2LifecycleStatus;
};

const v1Statuses = ["submitted", "triaged", "in_progress", "awaiting_customer", "resolved", "closed"] as const;
const v1DepartmentIds = ["transfer_payment", "account_support", "card_atm", "fraud_security", "loan_credit", "general_support"] as const;

export const v2DepartmentLabels: Record<V2DepartmentId, Record<Locale, string>> = {
  account_branch_services: { en: "Account & Branch Services", my: "အကောင့်နှင့် ဘဏ်ခွဲဝန်ဆောင်မှု" },
  cards_atm_pos: { en: "Cards, ATM & POS", my: "ကတ်၊ ATM နှင့် POS" },
  mobile_internet_banking: { en: "Mobile & Internet Banking", my: "မိုဘိုင်းနှင့် အင်တာနက်ဘဏ်လုပ်ငန်း" },
  transfers_payments_remittance: { en: "Transfers, Payments & Remittance", my: "ငွေလွှဲ၊ ပေးချေမှုနှင့် ငွေလွှဲပို့မှု" },
  loans_credit: { en: "Loans & Credit", my: "ချေးငွေနှင့် အကြွေး" },
  fraud_scam_unauthorized: { en: "Fraud, Scam & Unauthorized Transactions", my: "လိမ်လည်မှု၊ လှည့်ဖြားမှုနှင့် ခွင့်မပြုသော ငွေလွှဲမှု" },
  kyc_verification_restrictions: { en: "KYC, Verification & Account Restrictions", my: "KYC၊ အတည်ပြုခြင်းနှင့် အကောင့်ကန့်သတ်ချက်များ" },
  general_complaints: { en: "Customer Complaint Unit", my: "ဖောက်သည်တိုင်ကြားချက်များဌာန" },
};

export const v2LifecycleLabels: Record<V2LifecycleStatus, Record<Locale, string>> = {
  received: { en: "Received", my: "လက်ခံရရှိပြီး" },
  queued: { en: "In queue", my: "တန်းစီနေသည်" },
  assigned: { en: "Assigned to support", my: "အကူအညီပေးသူထံ ခန့်အပ်ထားသည်" },
  under_investigation: { en: "Under investigation", my: "စိစစ်နေသည်" },
  additional_information_required: { en: "More information needed", my: "အချက်အလက် ထပ်မံလိုအပ်သည်" },
  action_in_progress: { en: "Action in progress", my: "ဆောင်ရွက်နေသည်" },
  resolved: { en: "Resolution provided", my: "ဖြေရှင်းချက် ပေးပြီး" },
  closed: { en: "Closed", my: "ပိတ်သိမ်းပြီး" },
  reopened: { en: "Reopened", my: "ပြန်လည်ဖွင့်ထားသည်" },
};

export const staffEmploymentLabels: Record<StaffEmploymentState, Record<Locale, string>> = {
  invited: { en: "Invited", my: "ဖိတ်ကြားထားသည်" },
  active: { en: "Active", my: "တာဝန်ထမ်းဆောင်နေသည်" },
  on_leave: { en: "On leave", my: "ခွင့်ယူထားသည်" },
  suspended: { en: "Suspended", my: "ခေတ္တရပ်ဆိုင်းထားသည်" },
  departed: { en: "Departed", my: "အလုပ်မှထွက်ခွာပြီး" },
  archived: { en: "Archived", my: "မှတ်တမ်းသိမ်းထားသည်" },
};

export const staffAvailabilityLabels: Record<StaffAvailabilityState, Record<Locale, string>> = {
  available: { en: "Available", my: "ရရှိနိုင်သည်" },
  busy: { en: "Busy", my: "အလုပ်များနေသည်" },
  offline: { en: "Offline", my: "အော့ဖ်လိုင်း" },
  unavailable: { en: "Unavailable", my: "မရရှိနိုင်ပါ" },
};

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

function hasRawKey(value: Record<string, unknown>): boolean {
  return Object.keys(value).some((key) => key.includes("_"));
}

export function isV2DepartmentId(value: string | null | undefined): value is V2DepartmentId {
  return typeof value === "string" && v2DepartmentIds.includes(value as V2DepartmentId);
}

export function isV2LifecycleStatus(value: string | null | undefined): value is V2LifecycleStatus {
  return typeof value === "string" && v2LifecycleStatuses.includes(value as V2LifecycleStatus);
}

export function isStaffEmploymentState(value: string | null | undefined): value is StaffEmploymentState {
  return typeof value === "string" && staffEmploymentStates.includes(value as StaffEmploymentState);
}

export function isStaffAvailabilityState(value: string | null | undefined): value is StaffAvailabilityState {
  return typeof value === "string" && staffAvailabilityStates.includes(value as StaffAvailabilityState);
}

export function readTicketVersion(value: unknown): TicketVersion {
  if (!isRecord(value) || hasRawKey(value)) throw new Error("ticket version metadata is not safely readable");
  const keys = ["schemaVersion", "workflowVersion", "taxonomyVersion"] as const;
  const present = keys.map((key) => Object.prototype.hasOwnProperty.call(value, key));
  if (!present.some(Boolean)) {
    const status = value.status;
    const department = value.departmentId;
    if (typeof status !== "string" || !v1Statuses.includes(status as (typeof v1Statuses)[number])) {
      throw new Error("missing version metadata is not proven V1");
    }
    if (department !== null && department !== undefined && (typeof department !== "string" || !v1DepartmentIds.includes(department as (typeof v1DepartmentIds)[number]))) {
      throw new Error("missing version metadata is not proven V1");
    }
    return "v1";
  }
  if (!present.every(Boolean)) throw new Error("version metadata is incomplete");
  const versions = keys.map((key) => value[key]);
  if (versions.every((version) => version === "v1")) return "v1";
  if (versions.every((version) => version === "v2")) return "v2";
  throw new Error("ticket version is unknown or inconsistent");
}

export function getV2DepartmentLabel(departmentId: V2DepartmentId, locale: Locale): string {
  return v2DepartmentLabels[departmentId][locale];
}

export function getV2LifecycleLabel(status: V2LifecycleStatus, locale: Locale): string {
  return v2LifecycleLabels[status][locale];
}

export function parseV2TicketContract(value: unknown): V2TicketContract {
  if (!isRecord(value) || hasRawKey(value)) throw new Error("V2 ticket contains unsafe keys");
  const allowed = ["schemaVersion", "workflowVersion", "taxonomyVersion", "categoryId", "status"];
  if (Object.keys(value).some((key) => !allowed.includes(key))) throw new Error("V2 ticket contains unknown fields");
  if (value.schemaVersion !== "v2" || value.workflowVersion !== "v2" || value.taxonomyVersion !== "v2") {
    throw new Error("V2 ticket version is unknown or inconsistent");
  }
  if (!isV2DepartmentId(typeof value.categoryId === "string" ? value.categoryId : null)) throw new Error("unknown V2 category");
  if (!isV2LifecycleStatus(typeof value.status === "string" ? value.status : null)) throw new Error("unknown V2 lifecycle status");
  return value as V2TicketContract;
}
